spec/requests/new_survey_response_spec.rb

```diff
@@ -0,0 +1,106 @@
+require 'rails_helper'
+
+describe "New Survey Response" do
+  before do
+    allow(Broadcast).to receive(:emit)
+    allow(ActionsCache).to receive(:append)
+  end
+
+  let(:action) { Action.create(form_data: {}) }
+
+  let(:params) do
+    {
+      type: 'new_survey_response',
+      params: data,
+      meta: {
+        foo: 'bar',
+        action_id: action.id
+      }
+    }
+  end
+
+  let(:data) do
+    {
+      page:         "some-page-slug",
+      name:         "Pablo José Francisco de María",
+      postal:       "W1",
+      address1:     "The Lodge",
+      address2:     "High Street",
+      city:         "London",
+      country:      "United Kingdom",
+      action_age:   "101",
+      action_foo:   "Foo",
+      action_bar:   "Bar",
+      ignored:      "ignore me",
+      email:        "omar@sumofus.org",
+      source:       'FB',
+      akid:         '3.4234.fsdf'
+    }
+  end
+
+  context "Given an action doesn't exist" do
+
+    before do
+      VCR.use_cassette("action_existing_page") do
+        post '/message', params
+      end
+    end
+
+    it 'responds successfully' do
+      expect(response).to be_success
+    end
+
+    it 'stores resource ID in champaign action' do
+      expect(action.reload.form_data['ak_resource_id']).to match(/rest\/v1\/petitionaction\//)
+    end
+
+    it 'stores the new action data in the ActionsRepository' do
+      ak_action = ActionRepository.get(action.id)
+      expect(ak_action[:page_ak_id]).to match(/rest\/v1\/petitionpage\//)
+      expect(ak_action[:ak_id]).to be_present
+    end
+  end
+
+  context "Given the action already exists" do
+    let(:update_params) do
+      params.clone.tap do |p|
+        p[:params][:action_question_1] = '123'
+      end
+    end
+
+    before do
+      # Create action
+      VCR.use_cassette("action_existing_page") do
+        post '/message', params
+      end
+      expect(response.success?).to be_truthy
```

Comment:
I'm usually not a fan of this pattern neither. But I think in this case it's a nice thing to do to make sure the setup was done correctly. These specs are trying to test the update process for an action, hence they need the action to be created first. This `expect` makes sure that this setup was done correctly, otherwise if the setup fails it would do it silently making the failing spec for the update process more difficult to debug.